pkg/webhook/mutation.go

```diff
@@ -0,0 +1,184 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"context"
+	"flag"
+	"net/http"
+	"time"
+
+	"gomodules.xyz/jsonpatch/v2"
+
+	"github.com/open-policy-agent/cert-controller/pkg/rotator"
+	opa "github.com/open-policy-agent/frameworks/constraint/pkg/client"
+	"github.com/open-policy-agent/gatekeeper/apis"
+	"github.com/open-policy-agent/gatekeeper/pkg/controller/config/process"
+	"github.com/open-policy-agent/gatekeeper/pkg/util"
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	clientcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+type mutationResponse string
+
+// MutationEnabled indicates if the mutation feature is enabled
+var MutationEnabled *bool
+
+const (
+	mutationSkipResponse    mutationResponse = "skip"
+	mutationUnknownResponse mutationResponse = "unknown"
+)
+
+func init() {
+	MutationEnabled = flag.Bool("enable-mutation", false, "Enable the mutation webhook")
```

Comment:
nit: `--enable-mutation-webhook` (looking toward reconciliation loops in the future), also we should add that this flag is alpha to the docstring.

Also, does this need to be in init()? Can it be in var?

Finally, did we already commit a mutation toggle flag? I think I saw a should we just use that one for now?

https://github.com/open-policy-agent/gatekeeper/pull/930/files

(though it looks like I missed the fact that we don't need to put this in init() for that review)

Reply:
I rebased this on top of master, where we now have mutation.MutationEnabled.
imho one parameter for all of mutation should be enough. Do you think we should add a separate one for just the webhook?
We then have:
```
WebhookMutationEnabled = flag.Bool("enable-mutation-webhook", false, "Enable the mutation webhook")
....
if *mutation.MutationEnabled && WebhookMutationEnabled {
   // Enable mutation webhook
}
```